### PR TITLE
Look for node_modules in any recursive parent directory

### LINKED ISSUE #43: Support node recursive module resolution algorithm
Currently, cordova-fetch only checks the project's node_module directory, without checking parent directories. For example, in:

```
myproject/
  node_modules/
    dependency/ -> ../packages/dependency
  packages/
    dependency/
    subproject/
      index.js
      config.xml
      package.json
  package.json
```

If we `require('dependency')` from index.js, node will search through all ancestor directories for node_modules/dependency.

However, `cordova-fetch` doesn't support this. Instead, it will install `dependency` again (from the npm repository) inside myproject/packages/subproject/node_modules

This is a problem for use cases like yarn workspaces, where the *standard* node module resolution algorithm is used to help "dependency" be developed alongside "subproject".

Solving this problem could be a good way to improve support for yarn workspaces without needing to [add explicit support for yarn](https://github.com/apache/cordova-cli/pull/292#issuecomment-401331607).

### References:

- [resolve package](https://www.npmjs.com/package/resolve) - aims to implement node's algorithm fully, could be used as an alternative to get-installed-path
- the [node module resolution algorithm](https://nodejs.org/api/modules.html#modules_all_together)

### PR DESCRIPTION
**Pull request by ariofrio**
Resolves #43.

~This needs tests at the moment.~ Tests added.

### CODE REVIEW

**Comment by raphinesse** [comment]
I've added an implementation using `resolve`. It turns out that when using `resolve-from`, Node's `Module._pathCache` was causing more trouble than it was worth. Thanks for suggesting the package!

Do you think you could write a few tests for this, @ariofrio? That would be awesome! I think a few unit tests for `pathToInstalledPackage` with some local fixtures should be enough.

**Comment by ariofrio** [comment]
Thanks for working on this, @raphinesse! I'm pretty swamped right now; hopefully I can find some time for in a week or so to write a few tests for this.

**Comment by raphinesse** [comment]
@ariofrio FYI we're in the process of preparing the next major release. If we want this change in it, we should start work on it rather sooner than later. I'm sorry to say I can give no specific time frame though.

**Comment by ariofrio** [comment]
Working on this now. Not sure what a unit test that mocks `resolve` would accomplish, since most of the functionality that needs to be tested depends on the way `resolve` works.

I have in mind an integration test that creates this directory structure:

```
tmpdir/
  node_modules/
    dummy-local-plugin/
  nested-directory/
```

And then tries to run pathToInstalledPackage to find `dummy-local-plugin` from 'tmpdir' and then from 'tmpdir/nested-directory'.

What do you think?

**Comment by raphinesse** [comment]
The tests look good to me on a first glance :+1:

I never expected `resolve` to be mocked. How you laid out the tests was exactly how I imagined it.

One last thing: it would be nice to add a test that shows that globally installed modules will be taken into account too. Not sure how to best test this, but maybe you have an idea?

**Comment by raphinesse** [comment]
Or it might have been the linting errors that show up in the CI logs 😅

**Comment by ariofrio** [comment]
Huh, is taking into account globally installed modules expected behavior?

Seems like it would be error-prone, since running the same command on a different environment without that globally installed module would fail.

Not sure if resolve provides that functionality, anyway.

**Comment by raphinesse** [comment]
While I agree that taking global modules into account might be problematic, I would have thought `resolve` covers that functionality since it's part of the node module resolution algorithm.

Any way: all the more reason to assert the presence/absence of this behavior in the current implementation.

**Comment by ariofrio** [comment]
Good point about needing to test it either way.

At least on my environment, requiring a module that's installed globally doesn't work.

It seems that it's [not supposed to work](https://stackoverflow.com/a/15646750/237285), unless [NODE_PATH is set](https://nodejs.org/api/modules.html#modules_loading_from_the_global_folders) or the package is installed to a fixed set of three GLOBAL_FOLDERS (on Ubuntu using the nodesource distribution, globally installed modules don't go into any of them, but into /usr/lib/node_modules).

A grep through the source code seems to indicate that `resolve` doesn't read NODE_PATH or check the GLOBAL_FOLDERS, but expects them to be passed as the `paths` argument.

So to achieve full fidelity, we'd probably need to pass NODE_PATH and those three directories to resolve. Even then, in common cases globally installed modules won't be accessible.

Certainly doable. Looking for use-cases to work off of, it seems RisingStack and other [use NODE_PATH to avoid relative paths in require()](https://blog.risingstack.com/node-js-best-practices-part-2/), but they don't use it with whole packages.

---

Bottom line, globally installed modules *probably* shouldn't work, but it depends on whether the node environment is setup to install global packages to one of GLOBAL_FOLDERS. I don't know if we can assume that modern distributions avoid that.

Supporting NODE_PATH would make cordova-fetch fully compatible with npm resolution algorithm AFAIK, and the test for it seems simple.

**Comment by ariofrio** [comment]
Added NODE_PATH support and tests.

---

[This is the code](https://github.com/nodejs/node/blob/master/lib/internal/modules/cjs/loader.js#L789) that generates those GLOBAL_FOLDERS. It's straightforward to replicate this functionality in cordova-fetch, but the way it finds the node_prefix [has changed because the node binary moved](https://github.com/nodejs/node-v0.x-archive/blob/061151c/lib/module.js#L511), and I don't know if it could in the future.

Unfortunately, `globalPaths` seems to be undocumented, though it can be accessed through `require('module').globalPaths` in all node versions thus far, as far as I can tell. It can also be accessed by looking at the last three items of `module.path`.

All of these options have drawbacks, and supporting GLOBAL_FOLDERS would be "mostly for historic reasons". @raphinesse, up to you whether to support them and which option has the best trade-offs for the cordova project.

---

As far as testing global modules, I only see two ways of testing it:

1. actually install a global module and uninstall it afterwards, which could be disruptive and require sudo on some systems
2. use NODE_PATH to simulate global modules, but this is indistinguishable from testing NODE_PATH support

**Comment by raphinesse** [comment]
Great research @ariofrio! :medal_sports: 

I think supporting local lookup plus `NODE_PATH` lookup is just fine. No need for the hard-coded `GLOBAL_FOLDERS`. I will do a code review and then we need to run some tests with `cordova-lib` using this version.

**Comment by raphinesse** [comment]
So after *really* putting this to work to do dependency prefetching for the `cordova-lib` test suite, one problem popped up. I added a regression test and a fix for it.

File: index.js [hunk 1]
```diff
@@ -22,7 +22,7 @@ var events = require('cordova-common').events;
 var path = require('path');
 var fs = require('fs-extra');
 var CordovaError = require('cordova-common').CordovaError;
-const { getInstalledPath } = require('get-installed-path');
+const resolve = Q.denodeify(require('resolve'));
```
Comment (erisu) [inline index.js]: Not an issue but I know there was an initiative to remove the Q dependency and use Native promises. But this is no problem for now.

https://github.com/apache/cordova/issues/7

Comment (erisu) [reply index.js]: Everything looks good IMO.

**Comment by ariofrio** [comment]
Great! Thanks @raphinesse and @erisu for taking the time to work with me on this!

**Comment by raphinesse** [comment]
Thanks for your contribution @ariofrio!
